plugins/hydra_nevergrad_sweeper/hydra_plugins/hydra_nevergrad_sweeper/core.py

```diff
@@ -0,0 +1,228 @@
+        seed: Optional[int],
+        version: int,
+    ):
+        self.config: Optional[DictConfig] = None
+        self.launcher: Optional[Launcher] = None
+        assert version == 1, "Only version 1 of API is currently available"
+        self.job_results = None
+        self.optimizer = optimizer
+        self.noisy = noisy
+        self.budget = budget
+        self.num_workers = num_workers
+        self.seed = seed
+        self._direction = -1 if maximize else 1
+
+    def setup(
+        self,
+        config: DictConfig,
+        config_loader: ConfigLoader,
+        task_function: TaskFunction,
+    ) -> None:
+        self.config = config
+        self.launcher = Plugins.instantiate_launcher(
+            config=config, config_loader=config_loader, task_function=task_function
+        )
+
+    def sweep(self, arguments: List[str]) -> None:
+        assert self.config is not None
+        assert self.launcher is not None
+        # Construct the parametrization
+        params: Dict[str, Union[str, int, float, ng.p.Parameter]] = {}
+        for s in arguments:
+            key, value = s.split("=", 1)
+            params[key] = make_parameter(value)
+        parametrization = ng.p.Dict(**params)
+        parametrization.descriptors.deterministic_function = not self.noisy
+        parametrization.random_state.seed(self.seed)
+        # log and build the optimizer
+        name = "maximization" if self._direction == -1 else "minimization"
+        log.info(
+            "NevergradSweeper(optimizer=%s, budget=%s, num_workers=%s) %s",
+            self.optimizer,
+            self.budget,
+            self.num_workers,
+            name,
+        )
+        log.info("with parametrization %s", parametrization)
+        log.info("Sweep output dir : %s", self.config.hydra.sweep.dir)
+        optimizer = ng.optimizers.registry[self.optimizer](
+            parametrization, self.budget, self.num_workers
+        )
+        # loop!
+        remaining_budget = self.budget
+        all_returns: List[Any] = []
+        while remaining_budget > 0:
+            batch = min(self.num_workers, remaining_budget)
+            remaining_budget -= batch
+            candidates = [optimizer.ask() for _ in range(batch)]
+            overrides = list(
+                tuple(f"{x}={y}" for x, y in c.value.items()) for c in candidates
+            )
+            returns = self.launcher.launch(overrides)
+            # would have been nice to avoid waiting for all jobs to finish
+            # aka batch size Vs steady state (launching a new job whenever one is done)
+            for cand, ret in zip(candidates, returns):
+                optimizer.tell(cand, self._direction * ret.return_value)
+            all_returns.extend(returns)
+        recom = optimizer.provide_recommendation()
+        results_to_serialize = {"optimizer": "nevergrad", "nevergrad": recom.value}
+        results_to_serialize = json.loads(json.dumps(results_to_serialize))
```

Comment:
Why do we have to do this? I copied this from the `ax` sweeper and it says "to remove weird np values", but:
 1. I don't think there should be any here (then again I may be wrong)
 2. as far as I know, `json` has **never** been able to handle the numpy types.

Then again, if I don't add it, the conf cannot be saved... what am I missing?

Reply:
@jrapin, good question. I don't think it should be here.

@shagunsodhani, I thought we agreed to remove this and just add some logic to handle the serialization.

Reply:
@jrapin, can you describe the `optimization_result.yaml` layout?
At the moment there is no standard for this, but if we can agree on something both the Ax plugin and the NG plugin would implement it would be good.

@shagunsodhani, please describe what the Ax plugin is saving now.

Reply:
@omry The JSON serialization is "... to convert all the NumPy floats into python floats". Ax returns a dictionary with NumPy floats. It is possible that nevergrad returns floats and then this step is not required. We can add a custom serialization but it would just be typecasting NumPy floats to python floats using `float()`.

The proposed layout of `optimization_result.yaml` is 

`{
  "optimizer": <name of optimizer>, 
   <name of optimizer>: <value returned by the optimizer>}
`

For Ax, the returned value is [Tuple of parameterization and model predictions for it](https://ax.dev/api/service.html#ax.service.ax_client.AxClient.get_best_parameters)

@jrapin regarding "the conf cannot be saved", could you mention what error do you get? OmegaConf might not accept the custom type returned by nevergrad.

Reply:
@shagunsodhani, get more specific.
What is "value returned by optimizer" for Ax? what is it for Nevergrad?

Reply:
Using JSON as a kind of shortcut it for converting an object to OmegaConf is extremely lazy and also inefficient to boot.
If you think you are saving the need to loop on some floats ndarray, think again.

Reply:
Oh, that's funny. I guess this means np.float64 is an alias to float?

I nany case, I don't think we should rely on json here.
I would like to get some standardized output for the plugin output file (it we can also make it the return value from sweeper.sweep() ).

This can be a @dataclass compatible with OmegaConf structured configs, but the more important thing is to agree on what it's in it.

Initial proposal (fully made up, feel free to change).

```python
@dataclass
class SweeperResult
  name : str = MISSING
  best_parameters: Dict[str, Any] = MISSING # optimal parameters found
  best_achieved_result: float = MISSING     # optimal return value obtained
  sweeper_cfg : Any = MISSING               # The hydra.sweeper.params node
```
Makes more sense to do it in a followup diff.
@jrapin and @shagunsodhani, let's move the discussion on this the Hydra chat. will be less noisy.

Reply:
Chat sounds good to me. I am opening an issue - to make sure we do not forget about this. https://github.com/facebookresearch/hydra/issues/468